### ISSUE TITLE
medaka stitch OSError: error when opening file 'set_9.hdf`

### ISSUE THREAD

**Issue by jdcny24** [nanoporetech/medaka#338]
**Describe the bug**
I encountered an error (medaka.common.OverlapException) when trying to run `medaka stitch` on an older version of medaka and based on other threads, I figured that updating to the latest version 1.4.4 would resolve the issue. However, I am getting this error message which I am unable to find a solution for. 

> OSError: error when opening file 'set_9.hdf` 

The command I used was simply `time medaka stitch $HDF $OUT`

**Logging**
For the previous version 1.0.3 where my error message was medaka.common.OverlapException: 

>  The above exception was the direct cause of the following exception:
> 
> Traceback (most recent call last):
>   File "/cm/shared/apps/bio/medaka/conda/envs/medaka/bin/medaka", line 11, in <module>
>     sys.exit(main())
>   File "/cm/shared/apps/bio/medaka/conda/envs/medaka/lib/python3.6/site-packages/medaka/medaka.py", line 643, in main
>     args.func(args)
>   File "/cm/shared/apps/bio/medaka/conda/envs/medaka/lib/python3.6/site-packages/medaka/stitch.py", line 141, in stitch
>     for contigs in executor.map(worker, regions):
>   File "/cm/shared/apps/bio/medaka/conda/envs/medaka/lib/python3.6/concurrent/futures/process.py", line 366, in _chain_from_iterable_of_lists
>     for element in iterable:
>   File "/cm/shared/apps/bio/medaka/conda/envs/medaka/lib/python3.6/concurrent/futures/_base.py", line 586, in result_iterator
>     yield fs.pop().result()
>   File "/cm/shared/apps/bio/medaka/conda/envs/medaka/lib/python3.6/concurrent/futures/_base.py", line 432, in result
>     return self.__get_result()
>   File "/cm/shared/apps/bio/medaka/conda/envs/medaka/lib/python3.6/concurrent/futures/_base.py", line 384, in __get_result
>     raise self._exception
> medaka.common.OverlapException: Could not find viable junction for 66:8991086.1566-8991092.135 and 66:8991090.637-8991096.1686

Where the version is 1.4.4 and produces an OSError message:

> [E::fai_build_core] Format error, unexpected character at line 1
> Traceback (most recent call last):
>   File "/cm/shared/apps/bio/medaka/conda/envs/1.4.4/bin/medaka", line 11, in <module>
>     sys.exit(main())
>   File "/cm/shared/apps/bio/medaka/conda/envs/1.4.4/lib/python3.8/site-packages/medaka/medaka.py", line 707, in main
>     args.func(args)
>   File "/cm/shared/apps/bio/medaka/conda/envs/1.4.4/lib/python3.8/site-packages/medaka/stitch.py", line 155, in stitch
>     draft = pysam.FastaFile(args.draft)
>   File "pysam/libcfaidx.pyx", line 123, in pysam.libcfaidx.FastaFile.__cinit__
>   File "pysam/libcfaidx.pyx", line 183, in pysam.libcfaidx.FastaFile._open
> OSError: error when opening file `hdf/set_9.hdf`

**Environment (if you do not have a GPU, write No GPU):**
 - Installation method: conda
 - OS: CentOS 7.7
 - medaka version: 1.4.4
 - GPU model: Nvidia Tesla V100-PCIE GPU card
 - CUDA version 11.1

**Comment by cjw85** [comment]
Hi @jdcny24,

For medaka v1.4.4 the `stitch` command takes three arguments:

```
$ medaka stitch
usage: medaka stitch [-h] [--debug | --quiet] [--regions REGIONS [REGIONS ...]] [--threads THREADS] [--no-fillgaps] inputs [inputs ...] draft output
medaka stitch: error: the following arguments are required: inputs, draft, output
```

The `draft` argument is required after many users expressed a desire that unpolished regions should be included in the final output. I'm not sure how your command ran as with only two arguments you will receive:

```
$ medaka stitch thing.hdf output.fasta
usage: medaka stitch [-h] [--debug | --quiet] [--regions REGIONS [REGIONS ...]] [--threads THREADS] [--no-fillgaps] inputs [inputs ...] draft output
medaka stitch: error: the following arguments are required: output
```

I suspect one you have straightened-out this error you may see a similar error with v1.4.4 as you did with v1.0.3. Can you provide details how your HDF input was constructed, from what arguments and the nature of your data? The error:

```
medaka.common.OverlapException: Could not find viable junction for 66:8991086.1566-8991092.135 and 66:8991090.637-8991096.1686
```

Indicates two things to me:
1) the code has fallen back on a heuristic to piece together blocks of data rather than its more simple well defined approach. This heuristic was written for an experimental version of medaka and should not ordinarily be called.
2) You have some very large insertions in your reads compared to the reference to which they are aligned. The coordinates such as `8991086.1566` and `8991096.1686` are indicating 1566 and 1686 insertions columns in the pileup of reads following the reference positions `8991086` and `8991096` respectively.

**Comment by jdcny24** [comment]
hi @cjw85, 

i re-ran `medaka stitch $HDF $DRAFT $OUT` and got the same error message as the previous version, as you had expected (medaka.common.OverlapException: Overlaps are not equal in structure). i have 10 human samples, of which i encountered this error for 2 of them. the other 8 worked perfectly fine. after de novo assembly with shasta, i performed 3 iterations of minimap2/racon. subsequently, i did the following to generate the hdf files: 

> time mini_align -x -i bonito.fastq -r assembly-racon.fasta -P -m -p sample.calls_to_draft.bam -t 140
> --- in-house perl script to split the sam file into chunks ---
> time medaka consensus sample.calls_to_draft.bam *.hdf --model r941_prom_high_g360 --batch 500 --threads 8 --region chunks.txt

**Comment by cjw85** [comment]
To process in parallel multiple regions from a BAM it is not necessary to split the BAM into chunks, in fact this could well be the cause depending on what exactly your perl script is doing.

You can pass a list of regions to process, as you might to samtools:

```
medaka consensus input.bam output.hdf --regions contig1 contig2 ... contig759:1000-2000
```

**Comment by jdcny24** [comment]
apologies, i did not describe the splitting into chunks correctly. the perl script essentially just extracts the contigs and splits them into chunks, so my chunks.txt looks like this: 

> set_1:0 2 4 6 8 10 12 14 16 18 20 22 24 26 28 30 32 34 36 38
> set_2:40 42 44 46 48 50 52 54 56 58 60 62 64 66 68 70 72 74 76 78

then it would generate set_1.hdf set_2.hdf and so on. are you suggesting that i try to not do them in parallel?

**Comment by cjw85** [comment]
That might be OK. There would certainly be an issue if you were creating sub-contig splits in your data and not ensuring all reads overlapping sub-contig chunks were included in all chunks (with duplication).

Regardless of that, splitting the BAM is not necessary as you can simply submit your job array based around the `--regions` option as above.

I think you may want to inspect the BAM in a viewer like IGV to see whats going on with the alignments around the areas where medaka is struggling: as before the errors its throwing suggest some rather large insertions in reads relative to your contigs.

**Comment by jdcny24** [comment]
ah, i see. i will try to visualise the bam. if it is indeed an issue with large insertions in reads, do you have any suggestions on how to resolve this issue?

**Comment by femiliani** [comment]
I'm having a different but also stitching issue. I upgraded to the new medaka to access the Q20 models and since then i cannot get medaka stitch to run (either independently or as part of medaka_consensus). It always gives the following error:
```
This is medaka 1.4.4
Program    Version    Required   Pass
bcftools   1.12       1.11       True
bgzip      1.12       1.11       True
minimap2   2.22       2.11       True
samtools   1.12       1.11       True
tabix      1.12       1.11       True
Warning: Output medaka already exists, may use old results.
Not aligning basecalls to draft, calls_to_draft.bam exists.
Not running medaka consensus, consensus_probs.hdf exists.
[11:56:26 - DataIndx] Loaded 1/1 (100.00%) sample files.
[11:56:26 - DataIndx] Loaded 1/1 (100.00%) sample files.
Traceback (most recent call last):
  File "/home/f002sd4/miniconda3/envs/medaka/bin/medaka", line 11, in <module>
    sys.exit(main())
  File "/home/f002sd4/miniconda3/envs/medaka/lib/python3.8/site-packages/medaka/medaka.py", line 707, in main
    args.func(args)
  File "/home/f002sd4/miniconda3/envs/medaka/lib/python3.8/site-packages/medaka/stitch.py", line 195, in stitch
    contigs, gt = fill_gaps(contigs, args.draft)
  File "/home/f002sd4/miniconda3/envs/medaka/lib/python3.8/site-packages/medaka/stitch.py", line 120, in fill_gaps
    (ref_name, 0, contig_lengths[ref_name]), pieces))
KeyError: b'utg000001c'
Failed to stitch consensus chunks.
```

What's unusual is that when i try on a separate machine in a new conda environment it works just fine. Interestingly when i install on the problematic machine (even if i install into a new conda environment) it initially gives me an error that the package chardet is missing (below) and i need to install chardet to even get to the above error. But on the other machine it never gives me the chardet error. Comparing conda list between the two fresh installs they are identical.

**In the process of trouble shooting i tried moving the hdf file created with the working node to the broken node, and it fails to stitch. But moving the hdf from the broken to the new, the new is able to stitch it. So it's isolated to the stitching step.** 

I am new to computational work so if you have a suggestion about what may be going wrong/how to fix it (I have tried googling to no avail) it would be greatly appreciated.  Thank you very much! 

```
Traceback (most recent call last):
  File "/home/f002sd4/miniconda3/envs/medaka/bin/medaka", line 7, in <module>
    from medaka.medaka import main
  File "/home/f002sd4/miniconda3/envs/medaka/lib/python3.8/site-packages/medaka/medaka.py", line 11, in <module>
    import medaka.models
  File "/home/f002sd4/miniconda3/envs/medaka/lib/python3.8/site-packages/medaka/models.py", line 7, in <module>
    import requests
  File "/home/f002sd4/.local/lib/python3.8/site-packages/requests/__init__.py", line 44, in <module>
    import chardet
ModuleNotFoundError: No module named 'chardet'
```

**Comment by cjw85** [comment]
Hi @femiliani,

Please avoid posting on open issues with different symptoms. It is better to open a new issue.

In your case, I believe the forthcoming v1.5.0 will fix the `KeyError: b'utg000001c'` issue you are having. This is currently being built on bioconda.

**Comment by femiliani** [comment]
Hello @cjw85
Apologies for the faux pas. Thanks for the quick response, I look forward to 1.5. 

Have a good day.
